malariagen_data/ag3.py

```diff
@@ -341,3 +347,100 @@ def _results_cache_add_analysis_params(self, params):
         super()._results_cache_add_analysis_params(params)
         # override parent class to add AIM analysis
         params["aim_analysis"] = self._aim_analysis
+
+    def load_inversion_tags(self, inversion: str):
+        # needs to be modified depending on where we are hosting
+        path = "malariagen_data/resources/karyotype_tag_snps.csv"
+        df_tag_snps = pd.read_csv(path, sep=",")
+        return df_tag_snps.query(f"inversion == '{inversion}'").reset_index()
+
+    def karyotype_tags_n_alt(self, gt, alts, inversion_alts):
+        # could be Numba'd for speed but was already quick (not many inversion tag snps)
+        n_sites = gt.shape[0]
+        n_samples = gt.shape[1]
+
+        # create empty array
+        inv_n_alt = np.empty((n_sites, n_samples), dtype=np.int8)
+
+        # for every site
+        for i in range(n_sites):
+            # find the index of the correct tag snp allele
+            tagsnp_index = np.where(alts[i] == inversion_alts[i])[0]
+
+            for j in range(n_samples):
+                # count alleles which == tag snp allele and store
+                n_tag_alleles = np.sum(gt[i, j] == tagsnp_index[0])
+                inv_n_alt[i, j] = n_tag_alleles
+
+        return inv_n_alt
+
+    @check_types
+    def karyotype(
+        self,
+        inversion,
+        sample_sets: Optional[base_params.sample_sets] = None,
+        sample_query: Optional[base_params.sample_query] = None,
+    ) -> pd.DataFrame:
+        """
+        Infer karyotype from tag SNPs for a given inversion.
+
+        Parameters
+        ----------
+        inversion : str
+            Name of inversion to infer karyotype for. One of ["2La", "2Rb", "2Rc_gam", "2Rc_col", "2Rd", "2Rj"].
+        sample_sets : list of str, optional
+            Sample sets to include in the analysis. If None (default), use all sample sets.
+        sample_query : str, optional
+            Query to filter samples. If None (default), use all samples.
+        """
+
+        inversion_list = ["2La", "2Rb", "2Rc_gam", "2Rc_col", "2Rd", "2Rj"]
+        assert (
+            inversion in inversion_list
+        ), f"Inversion {inversion} not found - please select one of {inversion_list}"
```

Comment:
You could define a type for this, e.g.:

```
inversion_param: TypeAlias = Literal["2La", "2Rb", "2Rc_gam", "2Rc_col", "2Rd", "2Rj"]
```

...and then declare this as the type of the `inversion` param, in which case the validation will be done automatically by the `check_types` wrapper.

Reply:
Cool. Is it appropriate to put karyo_params.py in the anoph/ directory given that the karyo functions are in ag3? Or should I just define this inversion param in ag3.py 

also, I assume I was right not to put the code in anoph/ originally - i did that as it was ag3 only and not af1.

Reply:
Yeah this is Ag specific so I think it makes sense to put everything in the ag3.py module.